### PR TITLE
Chat integration

### PR DESCRIPTION
**Pull request by liuxess**
Created Chat CRUD and websockets for communication

### CODE REVIEW

**Review by ValdasPiekus — APPROVED** [review]
Reviewed

File: main/Repositories/Database/ChatRepository.cs [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+            _context.ChatMessages.Add(chatMessage);
+            CheckEntityAuditing();
+            await _context.SaveChangesAsync();
+        }
+
+        public async Task<Chat> GetChatFromParticipantsAsync(List<string> allParticipants)
+        {
+            var a = from c in _context.Chats
+                                .Where(c => c.Participants.Any() &&
+                                         c.Participants.All(p => allParticipants.Contains(p.UserId))
```
Comment (ValdasPiekus) [inline main/Repositories/Database/ChatRepository.cs:49]: Would there be any performance increase if you would have checked if allParticipants contain items, and if there are none - skipping the query and returning empty result?

File: main/Repositories/Database/ChatRepository.cs [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        public async Task<IEnumerable<Chat>> GetChatsWithLastMessageAndParticipantsAsync(string userId, int pageSize, DateTime? createdSince)
+        {
+            UserContext userContext = _contextAccessor.HttpContext.GetUserContext();
+            var a = from c in _context.Chats
+                                .Include(c=>c.Messages.OrderByDescending(m => m.CreatedOn).Take(1))
+                                .Include(c=>c.Participants)
+                                .Take(pageSize)
+                                .Where(c=>
+                                    c.Participants.Select(p=>p.UserId).Contains(userContext.UserID)
+                                    && (createdSince.HasValue && c.CreatedOn > createdSince)
```
Comment (ValdasPiekus) [inline main/Repositories/Database/ChatRepository.cs:65]: createdSince as a parameter can be null, will everything be okay in "c.CreatedOn > createdSince" check if it is null?

File: main/Repositories/Database/ChatRepository.cs [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+        public async Task<Chat> GetChatMessagesWithParticipantsDetails(long chatId, int size, DateTime? createdSince)
+        {
+            UserContext userContext = _contextAccessor.HttpContext.GetUserContext();
+            var a = _context.Chats
+                        .Include(c => c.Participants)
+                            .ThenInclude(p => p.User)
+                        .Include(c => c.Messages
+                                        .Take(size)
+                                        .Where(m => m.CreatedOn > createdSince)
```
Comment (ValdasPiekus) [inline main/Repositories/Database/ChatRepository.cs]: The same question with createdSince here too

File: main/Repositories/Database/ChatRepository.cs (hunk 2 above)
Comment (liuxess) [reply main/Repositories/Database/ChatRepository.cs:65]: createdSince.HasValue verifies that it has value

File: main/Repositories/Database/ChatRepository.cs (hunk 3 above)
Comment (liuxess) [reply main/Repositories/Database/ChatRepository.cs]: Added createdSince.HasValue

File: main/Repositories/Database/ChatRepository.cs (hunk 1 above)
Comment (liuxess) [reply main/Repositories/Database/ChatRepository.cs:49]: Returning assumed null if empty; good catch

**Pull request merged**